src/main/kotlin/org/jitsi/nlj/util/PacketCache.kt

```diff
@@ -16,85 +16,129 @@
+        val SIZE_PACKETS: String = "${PacketCache::class.java}.SIZE_PACKETS"
         private val defaultConfiguration = Configuration()
+
         init {
-            defaultConfiguration[NACK_CACHE_SIZE_MILLIS] = 1000
-            defaultConfiguration[NACK_CACHE_SIZE_PACKETS] = 500
+            defaultConfiguration[SIZE_PACKETS] = 500
         }
     }
 
     init {
-        timeout = Duration.ofMillis(defaultConfiguration.getInt(NACK_CACHE_SIZE_MILLIS).toLong())
-        maxNumElements = defaultConfiguration.getInt(NACK_CACHE_SIZE_PACKETS)
+        size = defaultConfiguration.getInt(SIZE_PACKETS)
     }
 
     private fun getCache(ssrc: Long): RtpPacketCache {
-        return packetCaches.computeIfAbsent(ssrc) { _ -> RtpPacketCache(timeout, maxNumElements) }
+        return packetCaches.computeIfAbsent(ssrc) { RtpPacketCache(size) }
     }
 
-    fun cachePacket(packet: RtpPacket) = getCache(packet.ssrc).insert(packet)
+    /**
+     * Stores a copy of the given packet in the cache.
+     */
+    fun insert(packet: RtpPacket) =
+        packetPredicate(packet) && getCache(packet.ssrc).insert(packet)
 
-    fun getPacket(ssrc: Long, seqNum: Int): RtpPacket? = getCache(ssrc).get(seqNum)
+    /**
+     * Gets a copy of the packet in the cache with the given SSRC and sequence number, if the cache contains it.
+     * The instance is wrapped in a [ArrayCache.Container].
+     */
+    fun get(ssrc: Long, seqNum: Int): ArrayCache<RtpPacket>.Container? = getCache(ssrc).get(seqNum)
 
+    /**
+     * Gets copies of the latest packets in the cache. Tries to get at least [numBytes] bytes total.
+     */
     fun getMany(ssrc: Long, numBytes: Int): Set<RtpPacket> = getCache(ssrc).getMany(numBytes)
 
     /**
-     * A cache of packets which maps their RFC3711 index to the packet itself.
-     * Each insertion, all packets (in order of index) up to the first one
-     * which is newer than [timeout] are pruned from the cache.
+     * Updates the timestamp of a packet in the cache (if it is in the cache). This is used when we re-transmit a
+     * packet in order to update the timestamp without re-adding the packet to the cache (which is expensive).
      */
-    private class RtpPacketCache(
-        timeout: Duration,
-        maxNumElements: Int
-    ) {
-        private val cache = TimeExpiringCache<Int, RtpPacket>(timeout, maxNumElements)
-        private val rfc3711IndexTracker = Rfc3711IndexTracker()
-
-        fun insert(packet: RtpPacket) {
-            val index = rfc3711IndexTracker.update(packet.sequenceNumber)
-            cache.insert(index, packet)
+    fun updateTimestamp(ssrc: Long, seqNum: Int, timeAdded: Long) = getCache(ssrc).updateTimestamp(seqNum, timeAdded)
+}
+
+/**
+ * Implements a cache for RTP packets.
+ */
+class RtpPacketCache(
+    size: Int,
+    synchronize: Boolean = true,
+    timeProvider: TimeProvider = TimeProvider()
+) : ArrayCache<RtpPacket>(size, synchronize, { it.clone() }, timeProvider) {
```

Comment:
is it possible to use a method reference instead of the lambda for clone?

Reply:
I'm not sure. If I use `RtpPacket::clone` would that work properly with subclasses? If you don't know I'll test tomorrow

Reply:
yeah, i just ran a test to double check